### ISSUE TITLE
C# Non-exported variable becomes null after build for Tool classes

### ISSUE THREAD

**Issue by discworld** [godotengine/godot#86959]
### Tested versions

- Reproducible in v4.2.1.stable.mono.official [b09f793f5]

### System information

Godot v4.2.1.stable.mono - Windows 10.0.22631 - GLES3 (Compatibility) - AMD Radeon(TM) Graphics (Advanced Micro Devices, Inc.; 31.0.21018.9001) - AMD Ryzen 5 PRO 5650U with Radeon Graphics (12 Threads)

### Issue description

On classes with [Tool] annotation, private variables can become null when recompiling the project. Manually changing exported properties in the editor fixes this, even if exported variable is unrelated to the private variable. Reloading the scene in the editor also fixes this.

The MRP below uses GetNode to fetch a node during _Ready. This works fine when loading the scene for the first time, however when changing the code and rebuilding, the private variable is null and _Ready is not called again.

### Steps to reproduce

- Load the MRP and open the "secondary.tscn" scene
- Output messages should read:
```
Ready
Target sprite is not null
End Ready
```
- Change the ParentNode.cs file, e.g. modify the comment on line 26
- Build project in Godot
- Output messages should have:
```
Target sprite is null
Target sprite is null
```
- Change "ShowTargetSprite" boolean from the "Node2D" inspector window
- Output messages should have:
```
Set ShowTargetSprite
Target sprite is not null
```

### Minimal reproduction project (MRP)

[LostState - MRP.zip](https://github.com/godotengine/godot/files/13859799/LostState.-.MRP.zip)

**Comment by bs-mwoerner** [comment]
When you rebuild the script, Godot saves the state of your current `ParentNode` object, loads the new code, creates a new `ParentNode` instance from the new code, and then restores the state (for fields and properties that are still there in the new code). 

This restoration of state necessarily happens in a certain order. In your example, Godot restores `ShowTargetSprite` first, then `_targetSprite`, then `_showTargetSprite`. Restoring `ShowTargetSprite` is done by calling your `set` accessor. This accessor prints the current state of `_targetSprite`. Since `ShowTargetSprite` is the first property to be restored, everything else still has its default value at this point, so your method prints `Target sprite is null`. Godot then goes on to restore `_targetSprite` and `_showTargetSprite`, but setting these doesn't print anything.

If you print the state of `_targetSprite` at any point after the state restoration has finished (such as when toggling `ShowTargetSprite`in the inspector), you will see the expected value.

As a general rule, relying on other properties to have certain values in code that may run in the middle of a deserialization isn't a good idea. There may be better ways to do this, but I like calling `IsInsideTree()` as a sanity check to prevent code from running on partially initialized objects.

**Comment by discworld** [comment]
Thank you for your reply. I wasn't aware of how Godot does the de-serialization for the private variables but your explanation makes sense. 
However, IsInsideTree() doesn't seem to help that much for Tool scrips. I've changed OnShowTargetSpriteChanged() as shown below:
```
private void OnShowTargetSpriteChanged()
{
    GD.Print("Show target sprite: " + _showTargetSprite);
    GD.Print("Target sprite" + (_targetSprite == null ? " is" : " is not") + " null");
    if (!IsInsideTree())
    {
        GD.Print("Not inside tree.");
    } else
    {
        GD.Print("Inside tree.");
    }
}
```

The output appears to show that setters are being called twice, last time with the actual value shown in the editor. But the private values are not yet restored by the time IsInsideTree() returns true. Is this intended as well?

```
Show target sprite: False
Target sprite is null
Not inside tree.
Show target sprite: True
Target sprite is null
Inside tree.
```

**Comment by bs-mwoerner** [comment]
Ah, true, apparently the object is added to the tree before the properties are restored, so `IsInsideTree()` makes sure that you can properly use `GetNode()` but won't guarantee that all properties have their final value.

If you really need to know when deserialization is complete (and I encourage you to question that, since it may indicate a problem with the software design), there is an `ISerializationListener` interface you can implement:

```C#
public partial class ParentNode : Node2D, ISerializationListener
{
    [...]
    public void OnBeforeSerialize()
    {

    }

    public void OnAfterDeserialize()
    {
        GD.Print("We're ready now. Target sprite" + (_targetSprite == null ? " is" : " is not") + " null");
    }
}
```

**Comment by discworld** [comment]
I agree that relying on ISerializationListener is not the best thing.

Looking around older issues, it seems something similar was discussed in the past for GDScript: https://github.com/godotengine/godot/issues/1697. A soft-reloading mechanism was implemented in that case. Achieving the same thing for C# would be great.

I'm not really sure what the solution is here but one thing that might improve things is not to restore the value of properties that have a custom set method, since those have a backing variable which is going to be set later on anyway.
